### PR TITLE
Fix S4200 FN: Add support for LibraryImportAttribute; run only on non-generated code

### PR DESCRIPTION
**Pull request by martin-strecker-sonarsource**
Part of #6488
Related to #6476

Rule S4200 should support LibraryImportAttribute.

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/NativeMethodsShouldBeWrappedTest.cs [hunk 1]
```diff
@@ -47,6 +47,7 @@ public class NativeMethodsShouldBeWrappedTest
                 .AddPaths("NativeMethodsShouldBeWrapped.CSharp11.cs")
                 .AddPaths("NativeMethodsShouldBeWrapped.CSharp11.SourceGenerator.cs")
                 .WithOptions(ParseOptionsHelper.FromCSharp11)
+                .WithConcurrentAnalysis(false)
```
Comment (andrei-epure-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Rules/NativeMethodsShouldBeWrappedTest.cs:50]: [education] why is this necessary?

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/NativeMethodsShouldBeWrapped.CSharp11.cs [hunk 2]
```diff
@@ -8,3 +8,32 @@ public interface IInterface
...
     public static extern virtual bool RemoveDirectory2(string name);  // Noncompliant
 }
+
+namespace LibraryImportAttributeTests
+{
+    // Copy the class to a .Net 7 project in VS and use "goto definition" on the methods to extract the code generated by the Microsoft.Interop.LibraryImportGenerator
+    public static partial class ExternMethods
+    {
+        [LibraryImport("foo.dll")]
+        public static partial void DllImportAttributeAppliedToThisFunction(); // Noncompliant
```
Comment (andrei-epure-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/NativeMethodsShouldBeWrapped.CSharp11.cs:18]: nitpick: should these be renamed `LibraryImportAttributeApplied...` instead of `DllImport`? because it's slightly confusing

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/NativeMethodsShouldBeWrapped.CSharp11.cs [hunk 3]
```diff
@@ -8,3 +8,32 @@ public interface IInterface
...
+
+        [LibraryImport("foo.dll", StringMarshalling = StringMarshalling.Utf8)]
+        public static partial void DllImportAttributeAppliedToGeneratedLocalFunction(string p); // Noncompliant
+
+        [LibraryImport("foo.dll", StringMarshalling = StringMarshalling.Utf8)]
+        private static partial void CompliantDllImportAttributeAppliedToGeneratedLocalFunction(string p); // Compliant
+
+        // Wrapper tests
+
+        public static void CompliantDllImportAttributeAppliedToThisFunctionWrapper(int i) // Noncompliant {{Make this wrapper for native method 'CompliantDllImportAttributeAppliedToThisFunction' less trivial.}}
```
Comment (andrei-epure-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/NativeMethodsShouldBeWrapped.CSharp11.cs:31]: nitpick: can you also add a test to ensure that when it has a "non trivial" wrapper, it passes (it's a bit of duplication in tests, but it can be a good regression test)

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/NativeMethodsShouldBeWrapped.CSharp11.SourceGenerator.cs [hunk 4]
```diff
@@ -0,0 +1,86 @@
+﻿// <auto-generated/>
+namespace LibraryImportAttributeTests
+{
```
Comment (andrei-epure-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/NativeMethodsShouldBeWrapped.CSharp11.SourceGenerator.cs:3]: can you also add a "trivial wrapper" here , to test `ReportTrivialWrappers` isn't triggered on generated code?

File: analyzers/src/SonarAnalyzer.CSharp/Rules/NativeMethodsShouldBeWrapped.cs [hunk 5]
```diff
@@ -34,20 +34,28 @@ public sealed class NativeMethodsShouldBeWrapped : SonarDiagnosticAnalyzer
         protected override void Initialize(SonarAnalysisContext context)
         {
             context.RegisterSymbolAction(ReportPublicExternalMethods, SymbolKind.Method);
-            context.RegisterSyntaxNodeAction(ReportTrivialWrappers, SyntaxKind.MethodDeclaration);
+            context.RegisterSyntaxNodeActionInNonGenerated(ReportTrivialWrappers, SyntaxKind.MethodDeclaration);
```
Comment (andrei-epure-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/NativeMethodsShouldBeWrapped.cs:37]: [education] why is this necessary to not be raised on generated code?

I get why it's necessary for the case of `LibraryImport` (to not raise the issue twice + users don't control the source generator) - is it the case that in general we should not raise on generated code ever?

**Review by andrei-epure-sonarsource — CHANGES_REQUESTED** [review]
This introduces a difference in the existing behavior of the rule --> please update the PR title description, as this will show up in the github history

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/NativeMethodsShouldBeWrappedTest.cs (hunk 1 above)
Comment (martin-strecker-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/Rules/NativeMethodsShouldBeWrappedTest.cs:50]: The `<autogenerated />` comment gets wrapped in the concurrent namespace and the logic to detect autogenerated code is no longer working. https://github.com/SonarSource/sonar-dotnet/pull/6211 should fix this.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/NativeMethodsShouldBeWrapped.CSharp11.cs (hunk 2 above)
Comment (martin-strecker-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/NativeMethodsShouldBeWrapped.CSharp11.cs:18]: No. The DllImport attribute gets applied by the LibraryImport source generator to this method. The actual method after the source generator run is the combination of the two partial definitions:
```cs
        [LibraryImport("foo.dll")]
        [System.Runtime.InteropServices.DllImportAttribute("foo.dll", EntryPoint = "DllImportAttributeAppliedToThisFunction", ExactSpelling = true)]
        public static extern partial void DllImportAttributeAppliedToThisFunction(); 
```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/NativeMethodsShouldBeWrapped.cs (hunk 5 above)
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/NativeMethodsShouldBeWrapped.cs:37]: Yes. It was requested by Pavel at one point that we should fix this here. I added it a item 4 on the todo list here https://github.com/SonarSource/sonar-dotnet/issues/6488 when he suggested it. It is also needed to properly support the LibraryImport attribute.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/NativeMethodsShouldBeWrapped.CSharp11.SourceGenerator.cs (hunk 4 above)
Comment (andrei-epure-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/NativeMethodsShouldBeWrapped.CSharp11.SourceGenerator.cs:3]: Someone might not be aware of your initial intention and manually change the file in the future.

What do you suggest to avoid this situation? Can this process be automated and part of the test execution?

Comment (martin-strecker-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/NativeMethodsShouldBeWrapped.CSharp11.SourceGenerator.cs:3]: > Can this process be automated and part of the test execution?

I wish we could but this is pretty difficult to achieve. The source generators are itself based on analyzers (The ImportLibraryGenerator.dll gets passed to csc.exe via the /analyzer switch). If we want to enable our test infrastructure to support this (which is desirable in the long run), we need to dedicate a sprint to it.

> Or add a comment at the top of the file, at least?

On top doesn't work because of the `<auto-generated>` comment. I was thinking about a comment too, but this violates the copy/paste idea also. if we add a comment, we need to add it to the test method. I could use my bullet point list from above as a template. What do you think about this?

Comment (andrei-epure-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/NativeMethodsShouldBeWrapped.CSharp11.SourceGenerator.cs:3]: yes, a comment above the test method sounds good to me

File: analyzers/src/SonarAnalyzer.CSharp/Rules/NativeMethodsShouldBeWrapped.cs (hunk 5 above)
Comment (andrei-epure-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/NativeMethodsShouldBeWrapped.cs:37]: the PR name should reflect that IMO because it's a change of functionality for the whole rule and it's not only related to adding support for a new attribute

**Review by andrei-epure-sonarsource — APPROVED** [review]
LGTM, I will update the PR name and put in the message the info about non generated detection change of functionality
